### ISSUE TITLE
Recognition of postcondition check in error message

### ISSUE THREAD

**Issue by PavelVozenilek** [ziglang/zig#1312]
I tried (via the online compiler) whether post-condition checks do work:

```
const assert = @import("std").debug.assert;

pub fn main() void {
    var x : i32  = 1;
    defer assert(x == 2);
}

```

and everything seems to be OK, in both debug and release. Except for error message when the check fails, which is :

```
reached unreachable code
[37;1m/opt/zig/lib/zig/std/debug/index.zig:86:13[0m: [2m0x203039 in ??? (.code)[0m
            unreachable; // assertion failure
            [32;1m^[0m
[37;1m/home/runner/.code.tio:7:17[0m: [2m0x2200be in ??? (.code)[0m
    defer assert(x == 2);
                [32;1m^[0m
[37;1m/opt/zig/lib/zig/std/special/bootstrap.zig:86:22[0m: [2m0x220089 in ??? (.code)[0m
            root.main();
                     [32;1m^[0m
[37;1m/opt/zig/lib/zig/std/special/bootstrap.zig:70:20[0m: [2m0x220029 in ??? (.code)[0m
    return callMain();
                   [32;1m^[0m
[37;1m/opt/zig/lib/zig/std/special/bootstrap.zig:64:39[0m: [2m0x21fd9f in ??? (.code)[0m
    std.os.posix.exit(callMainWithArgs(argc, argv, envp));
                                      [32;1m^[0m
[37;1m/opt/zig/lib/zig/std/special/bootstrap.zig:39:5[0m: [2m0x21fc40 in ??? (.code)[0m
    @noInlineCall(posixCallMainAndExit);
    [32;1m^[0m

```

Since this idiom (`defer assert(...)` ) could mean nothing else but post-condition check, perhaps the compiler could recognize this case and print usable error message.

-----------

Also, post-condition may be not valid if error is returned. There's `errdefer`,  but it is of no use here.  Something like `success-defer` would be handy, or perhaps, to avoid even more reserved words,  `defer/defer(err)/defer(!err)`.

**Comment by phase** [comment]
I believe he is wanting a compiler error/warning for `defer assert()`, but he then provides a use case _for_ `defer assert()`, a post condition check.

**Comment by PavelVozenilek** [comment]
@andrewrk: if compile sees code like:

```
fn foo()
{
   defer { 
      ... whatever
      assert(...); // top level assert == post-condition
      ...
      assert(...);
      ...
}

```

it modifies the asserts somehow. If that assert fires it would say something about failed post-condition, instead of the generic stack trace litany.


Futhermore, having something like `defer(!err)` (equivalent to `scope(success)` in D) would be desirable for post-conditions.

**Comment by thejoshwolfe** [comment]
> perhaps the compiler could recognize this case and print usable error message.

how was the error message unusable? it looks fine to me, except that it includes terminal control sequences.

**Comment by PavelVozenilek** [comment]
@thejoshwolfe: 

* The very first part of the message leaks out internal implementation (something, something about unreachable). 
* The most important part (which assertion failed) is hidden behind file/line/position and some other details.


Even the whole stack trace is over the top, for the most of the time. I did implement such functionality (win32 has support for extracting runtime stack traces from debug info). It sounded as great idea,  but after a while it got so annoying that I removed it and made it optional with a switch.
